non-square VAE tiling

wbruna opened the pull request:
Just an experiment to give us even more parameters to play with 🙂

I noticed sd_tiling didn't have anything _forcing_ the tiles to be square, so I've extended SD_TILE_SIZE to support specifying tile sizes directly (like "32x48"), or as a floating-point factor applied to the image dimensions.

The factor thing is interesting, because it can be used to keep the number of tiles in both directions more or less constant, which makes a specific value fast for many different image ratios. For instance, around 15.2 (weird number because I just apply it as-is right now) is just enough to keep it at 4 iterations, which on my card is consistently faster than non-tiled VAE. In hindsight, it probably makes more sense to specify it directly as a number of tiles in each direction, though.

wbruna committed:
refactor tile number calculation

wbruna committed:
support non-square tiles

wbruna committed:
add env var to change tile overlap

stduhpf commented on stable-diffusion.cpp:
```diff
@@ -1427,13 +1427,42 @@ class StableDiffusionGGML {
+            size_t x_pos = sd_tile_size_str.find('x');
             try {
-                tile_size = std::stoi(sd_tile_size_str);
+                if (dot_pos != std::string::npos) {
+                    float factor = std::stof(sd_tile_size_str);
+                    tile_size_x = W * (decode ? 8 : 1) / factor;
+                    tile_size_y = H * (decode ? 8 : 1) / factor;
+                }
```
I find this part confusing, it's not doing what I expected it to at first.

I have to problems with it: 

- the size of the tile is in units of latent space (it defaults to 32x32 latent pixels, so 256x256 rgb pixels). W and H are in the same latent space units when decoding, and in image rgb space units when encoding.  So I think it should rather be divided by 8 when not decoding rather than multiplied by 8 when decoding. 
- Shouldn't it be multiplied by the factor rather than divided by it? For example, when testing it, my intuition was to set `SD_TILE_SIZE` to 0.25 and I was expecting tiles that were 4x smaller than the latent image in each dimension, but I was very surprised to get tiles that were actually 32 times (8/0.25) bigger than the latent image in each dimension.

Also it would be cool to be able to combine the "." and the "x".

stduhpf commented:
Overall nice work, I like the way you refactored the `sd_tiling_calc_tiles()` thing.

wbruna commented:
I've fixed the latent/image confusion, and added support for two floating point factors for the env var.

About the factor: I don't think it makes sense to have more than 1 tile across a dimension, so I'm considering a factor below 1 as meaning "fraction of the latent dimension", and above 1 "how many tiles across the latent dimension", so we can try out both to see which one works best.

stduhpf commented:
> "how many tiles across the latent dimension"

That's not taking the overlap into account, right?

wbruna commented:
> That's not taking the overlap into account, right?

No, just a plain division right now.

stduhpf commented:
The way to compute the tile size so there are only `N` tiles in one dimension, with an overlap factor of `O` would be to set the factor (multiplicative) to something like `(1-O)/(N-O)`. If you want to add that, let me know, otherwise I will just merge this later today as it is.

wbruna committed:
add safeguards and better error messages for SD_TILE_OVERLAP

wbruna commented:
Hm... Just let me check if I'm understanding how that overlapping factor works.

We want N tiles, of size T, to fill the length L. To avoid boundary issues, we start by placing a first tile T. _If_ the overlapping factor means "which fraction of the tile is overlapping with its neighbor", each of the additional N-1 tiles will overlap T · overlap with the already placed tile, thus covering an additional area of T · (1-overlap). So we have:
```
L = T + (N-1)*T*(1-overlap)
T = L / ( 1 + (N-1)*(1-overlap) )
T = L / ( N - N*overlap + overlap )
```
For a 64 side with 3 tiles and overlap 0.5, this gives me a pretty reasonable tile size 32.

Am I on the right track here? 🙂

stduhpf commented:
Yes that seems right. I'm not sure how I ended up with the `(1-O)/(N-O)`.

wbruna committed:
add safeguards and include overlapping factor for SD_TILE_SIZE

wbruna commented:
It's _mostly_ working. I had to request slightly different sizes, depending on the overlapping factor; for instance, for a 768x768 image with overlapping 0.5, a 4.0 factor gives me 4 tiles across; lowering the overlapping to 0.3, I get 5; but 3.9 gives me the expected 4... so it could be just rounding.

wbruna commented:
> It's _mostly_ working. I had to request slightly different sizes, depending on the overlapping factor; for instance, for a 768x768 image with overlapping 0.5, a 4.0 factor gives me 4 tiles across; lowering the overlapping to 0.3, I get 5; but 3.9 gives me the expected 4... so it could be just rounding.

Indeed, adding std::round calls when multiplying by the factors seems to avoid that weirdness.

wbruna committed:
avoid rounding issues when specifying SD_TILE_SIZE as a factor

stduhpf commented on ggml_extend.hpp:
```diff
@@ -607,8 +607,38 @@ __STATIC_INLINE__ void ggml_tensor_scale_output(struct ggml_tensor* src) {
+    num_tiles_dim = (small_dim - tile_overlap) / non_tile_overlap;
+    int overshoot_dim = ((num_tiles_dim + 1) * non_tile_overlap + tile_overlap) % small_dim;
+
+    if ((overshoot_dim != non_tile_overlap) && (overshoot_dim <= num_tiles_dim * (tile_size / 2 - tile_overlap))) {
+        // if tiles don't fit perfectly using the desired overlap
+        // and there is enough room to squeeze an extra tile without overlap becoming >0.5
+        num_tiles_dim++;
+    }
```
@wbruna I think it's actually because of this. 
If I remember correctly I added this to make sure the overlap was preferably bigger rather than smaller than the target (because less overlap tend to cause more noticable transitions).

stduhpf replied:
If your fix with rounding doesn't work, removing these would work, though I think it's preferable to keep it.

stduhpf replied:
Seems to be working fine so far, no matter the overlap.

stduhpf commented on stable-diffusion.cpp:
```diff
@@ -1427,23 +1427,91 @@ class StableDiffusionGGML {
+                    LOG_WARN("SD_TILE_OVERLAP too low, setting it to 0.0");
+                    tile_overlap = 0.0;
+                }
+                else if (tile_overlap > 0.95) {
+                    LOG_WARN("SD_TILE_OVERLAP too high, setting it to 0.95");
+                    tile_overlap = 0.95;
+                }
```
Overlap > 0.5 isn't supported, because that would mean more than 2 tiles (per dim) would overlap at some points. And supporting that would make the code much more complicated for something that has no real benefit anyways.

wbruna committed:
lower SD_TILE_OVERLAP limit

stduhpf commented:
All good now i think
